File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        // 하드디스크는 한 번에 하나의 작업만 수행할 수 있음
        // 평균 시간을 줄이는 방법을 찾아라
        int answer = 0;
        int time = 0;
        int idx = 0;
        
        Arrays.sort(jobs, (o1,o2) -> o1[0] - o2[0]);
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1,o2) -> o1[1] - o2[1]);
        
        while(!pq.isEmpty() || idx < jobs.length) {
            // 현재 시간보다 요청시간이 작 or 같은  요청시점을 추가
            while(idx < jobs.length && time >= jobs[idx][0]) {
                pq.offer(jobs[idx++]);
            }
            
            // 현재 시간보다 요청시간이 작거나 같은 작업이 없으면 제일 요청시점 빠른애 추가 
            if(pq.isEmpty()) {
                time = jobs[idx][0];
            }
            
            // 큐에 작업 존재 -> 제일 빨리 끝나는애부터 처리
            else {
                int[] job = pq.poll();
                time += job[1];
                answer += time - job[0];
            }
        }
        
        return answer / jobs.length;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int timer = 0;

bool compare(vector<int>&a, vector<int>& b){
    return a[1] <= b[1];
}

int solution(vector<vector<int>> jobs) {
    int answer = 0;
    int cnt = jobs.size();
    sort(jobs.begin(), jobs.end() , compare);
    
    while(!jobs.empty()){
        for(int i=0; i<jobs.size(); i++){
            // 타이머 >= 업무 요청 시간 
            if(timer >= jobs[i][0]){
                // 업무 걸리는 시간 +
                timer += jobs[i][1];
                answer += timer - jobs[i][0];
                jobs.erase(jobs.begin() + i);
                break;
            }
            if(i == jobs.size() - 1) timer++;
        }
    }
    
    return answer/cnt;
}